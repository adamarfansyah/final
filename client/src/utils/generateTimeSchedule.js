import moment from 'moment';

const generateTimeSlots = (date, start, end, intervalMinutes) => {
  const timeSlots = [];
  const currentTime = moment(start);

  while (currentTime.isBefore(end)) {
    const startTime = currentTime.format('HH:mm');
    currentTime.add(intervalMinutes, 'minutes');
    const endTime = currentTime.format('HH:mm');

    timeSlots.push({ start: `${date}T${startTime}`, end: `${date}T${endTime}` });
  }

  return timeSlots;
};

const renderDate = (date) => {
  const dateAdjusted = moment(date);
  const dateFormat = dateAdjusted.date();
  const monthFormat = dateAdjusted.format('MMMM');
  const yearFormat = dateAdjusted.year();

  return `${dateFormat} ${monthFormat} ${yearFormat}`;
};

const formattedTimeCardSchedule = (time) => {
  const formattedStartTime = moment(time.start).format('hh:mm A');
  const formattedEndTime = moment(time.end).format('hh:mm A');

  return `${formattedStartTime} - ${formattedEndTime}`;
};

export { renderDate, generateTimeSlots, formattedTimeCardSchedule };
