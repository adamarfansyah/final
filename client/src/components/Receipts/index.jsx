import PropTypes from 'prop-types';

import formattedNumber from '@utils/formattedNumber';
import { formattedTimeCardSchedule, renderDate } from '@utils/generateTimeSchedule';

import classes from './style.module.scss';

const Receipts = ({ transactionDetail }) => (
  <div className={classes.transactionDetail} data-testid="Receipts">
    {Object.keys(transactionDetail).length === 0 ? (
      <div>Something went wrong</div>
    ) : (
      <div className={classes.receipt}>
        <div className={classes.title}>Receipt</div>
        <div className={classes.wrapper}>
          <div className={classes.orderId} data-testid="ReceiptTitle">
            {transactionDetail.orderId}
          </div>
        </div>
        <div className={classes.date}>{renderDate(transactionDetail.createdAt)}</div>
        <div className={classes.contents}>
          <div className={classes.box}>
            <div className={classes.subTitle}>Order ID</div>
            <div className={classes.subDesc}>{transactionDetail.orderId}</div>
          </div>
          <div className={classes.box}>
            <div className={classes.subTitle}>Venue</div>
            <div className={classes.subDesc}>{transactionDetail?.BookedVenue?.name}</div>
          </div>
          <div className={classes.box}>
            <div className={classes.subTitle}>Customer Name</div>
            <div className={classes.subDesc}>
              {transactionDetail?.userBook?.firstName} {transactionDetail?.userBook?.lastName}
            </div>
          </div>
          <div className={classes.box}>
            <div className={classes.subTitle}>Merchant Name</div>
            <div className={classes.subDesc}>{transactionDetail?.merchantBook?.name}</div>
          </div>
        </div>
        <div className={classes.box}>
          <div className={classes.subTitle}>Booked Hours</div>
          <div className={classes.subDesc}>
            {formattedTimeCardSchedule({ start: transactionDetail.startBook, end: transactionDetail.endBook })}
          </div>
        </div>
        <div className={classes.grossAmount}>
          <div className={classes.grossAmountTitle}>Total</div>
          <div>{formattedNumber(transactionDetail.amount)}</div>
        </div>
      </div>
    )}
  </div>
);

Receipts.propTypes = {
  transactionDetail: PropTypes.object,
};

export default Receipts;
