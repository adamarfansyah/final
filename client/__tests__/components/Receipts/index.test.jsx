import { render } from '@testing-library/react';
import { formattedTimeCardSchedule } from '@utils/generateTimeSchedule';
import Receipts from '@components/Receipts';

const mockProps = {
  id: 1,
  BookedVenue: {
    name: 'Lapangan gbk 1',
  },
  userBook: {
    firstName: 'Adam Sake',
    lastName: 'Arfansyah',
  },
  merchantBook: {
    name: 'Gelora Bung Karno',
  },
  orderId: 'Lapangan GBK 1-1702866864-2',
  startBook: '2023-12-17T07:00',
  endBook: '2023-12-17T08:00',
  amount: '20000000',
  createdAt: '2023-12-17T13:54:52',
};

const wrapper = render(<Receipts transactionDetail={mockProps} />);

describe('Receipts Component', () => {
  test('renders correctly with transaction details', () => {
    const { getByText, getByTestId } = wrapper;
    expect(getByText(/Receipt/i)).toBeInTheDocument();
    expect(getByTestId('ReceiptTitle')).toBeInTheDocument();
    expect(getByText(mockProps.BookedVenue.name)).toBeInTheDocument();
    expect(getByText(`${mockProps.userBook.firstName} ${mockProps.userBook.lastName}`)).toBeInTheDocument();
    expect(getByText(mockProps.merchantBook.name)).toBeInTheDocument();
    expect(
      getByText(formattedTimeCardSchedule({ start: mockProps.startBook, end: mockProps.endBook }))
    ).toBeInTheDocument();
    expect(getByText('Rp 20.000.000,00')).toBeInTheDocument();
  });

  test('Should match with snapshot', () => {
    const { asFragment } = wrapper;
    expect(asFragment()).toMatchSnapshot();
  });
});
