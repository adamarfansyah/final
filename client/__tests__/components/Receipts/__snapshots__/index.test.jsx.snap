// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Receipts Component Should match with snapshot 1`] = `<DocumentFragment />`;
